js/src/logger.ts

```diff
@@ -1108,123 +1024,98 @@ export async function summarize(
+ * Return the currently-active span for logging (set by one of the `traced` methods). If there is no active span, returns a no-op span object, which supports the same interface as spans but does no logging.
  *
  * See `Span` for full details.
  */
 export function currentSpan(): Span {
-  return _state.currentSpan.getStore() ?? noopSpan;
+  return _state.currentSpan.getStore() ?? NOOP_SPAN;
 }
 
 /**
- * Toplevel function for starting a span. It checks the following (in precedence order):
- *  * Currently-active span
- *  * Currently-active experiment
- *  * Currently-active logger
- *
- * and creates a span in the first one that is active. If none of these are active, it returns a no-op span object.
- *
- * Unless a name is explicitly provided, the name of the span will be the name of the calling function, or "root" if no meaningful name can be determined.
- *
- * We recommend running spans within a callback (using `traced`) to automatically mark them as current and ensure they are terminated. If you wish to start a span outside a callback, be sure to terminate it with `span.end()`.
- *
- * See `Span.startSpan` for full details.
+ * Mainly for internal use. Return the parent object for starting a span in a global context.
  */
-export function startSpan(args?: StartSpanOptionalNameArgs): Span {
-  const { name: nameOpt, ...argsRest } = args ?? {};
-  const name =
-    (nameOpt ?? iso.getCallerLocation()?.caller_functionname) || "root";
+export function getSpanParentObject(): Span | Experiment | Logger {
   const parentSpan = currentSpan();
-
-  if (!parentSpan) {
-    throw new Error(
-      "Cannot call startSpan() from outside a trace. Please wrap this code in a traced() callback."
-    );
+  if (!Object.is(parentSpan, NOOP_SPAN)) {
+    return parentSpan;
   }
-
-  if (!Object.is(parentSpan, noopSpan)) {
-    return parentSpan.startSpan(name, argsRest);
-  }
-
   const experiment = currentExperiment();
   if (experiment) {
-    return experiment.startSpan({ name, ...argsRest });
+    return experiment;
   }
-
   const logger = currentLogger();
   if (logger) {
-    throw new Error(
-      "Cannot start a span within a logger from startSpan(). Use logger.startSpan() instead."
-    );
+    return logger;
   }
-
-  return noopSpan;
+  return NOOP_SPAN;
 }
 
 /**
- * Wrapper over `braintrust.startSpan`, which passes the initialized `Span` it to the given callback and ends it afterwards. See `Span.traced` for full details.
+ * Toplevel function for starting a span. It checks the following (in precedence order):
+ *  * Currently-active span
+ *  * Currently-active experiment
+ *  * Currently-active logger
+ *
+ * and creates a span under the first one that is active. If none of these are active, it returns a no-op span object.
+ *
+ * See `Span.traced` for full details.
  */
-export function traced<R>(
+export async function traced<R>(
```

Comment:
I know we've gone back & forth on this, but I'm mildly worried about making this function `async`, since I think it graph colors any uses of `traced` to also have to be `async` functions.

I know it's a lot to ask, but I _think_ we could solve this by making the `rootProject: RegisteredProject` arg in `SpanImpl`'s constructor a `Promise<RegisteredProject>`. Then, we can make the subfields of `objectInfo` also promises. And in the flush thread, we can check the values of the dictionary, and if are any are promises await them.

WDYT?

Reply:
Yeah I think pushing all the async work into the background logger could be a good design. As discussed offline, I separated out the python changes into https://github.com/braintrustdata/braintrust-sdk/pull/74, and we'll fix the JS API separately.